File: CollaborativeToDoApp.js
class Task {
  constructor(taskId, description, category, assignedUser) {
    this.taskId = taskId;
    this.description = description;
    this.category = category;
    this.assignedUser = assignedUser;
    this.completed = false;
    this.createdAt = new Date();
    this.updatedAt = new Date();
  }

  setDescription(newDescription) {
    this.description = newDescription;
    this.updatedAt = new Date();
  }

  setCategory(newCategory) {
    this.category = newCategory;
    this.updatedAt = new Date();
  }

  setCompleted(status) {
    this.completed = status;
    this.updatedAt = new Date();
  }

  getStatus() {
    return this.completed ? "Completed" : "Pending";
  }

  formatDate(date) {
    return date.toISOString().replace("T", " ").substring(0, 19);
  }

  toDetailedString() {
    return `
--------------------------------------------
Task ID     : ${this.taskId}
User        : ${this.assignedUser}
Description : ${this.description}
Category    : ${this.category}
Status      : ${this.getStatus()}
Created     : ${this.formatDate(this.createdAt)}
Updated     : ${this.formatDate(this.updatedAt)}
--------------------------------------------`;
  }

  toString() {
    return `Task[ID=${this.taskId}, User=${this.assignedUser}, Description=${this.description}, Category=${this.category}, Status=${this.getStatus()}]`;
  }
}

class User {
  constructor(username) {
    this.username = username;
    this.taskIds = [];
    this.createdAt = new Date();
  }

  addTaskId(taskId) {
    if (!this.taskIds.includes(taskId)) this.taskIds.push(taskId);
  }

  removeTaskId(taskId) {
    this.taskIds = this.taskIds.filter(id => id !== taskId);
  }

  hasTask(taskId) {
    return this.taskIds.includes(taskId);
  }

  getTaskCount() {
    return this.taskIds.length;
  }

  toString() {
    return `User[username=${this.username}, tasks=${this.taskIds.length}]`;
  }
}

class TaskManager {
  constructor() {
    this.tasks = new Map();
    this.users = new Map();
    this.taskCounter = 1;
  }

  async registerUser(username) {
    if (this.users.has(username)) {
      console.log(`User '${username}' already exists.`);
      return null;
    }
    const user = new User(username);
    this.users.set(username, user);
    console.log(`User '${username}' registered successfully.`);
    return user;
  }

  async createTask(username, description, category) {
    const user = this.users.get(username);
    if (!user) {
      console.log(`User '${username}' not found. Please register first.`);
      return null;
    }

    const taskId = `TASK-${String(this.taskCounter++).padStart(4, "0")}`;
    const task = new Task(taskId, description, category, username);
    this.tasks.set(taskId, task);
    user.addTaskId(taskId);

    console.log(`Task '${taskId}' created for user '${username}'.`);
    return task;
  }

  async updateTaskDescription(taskId, newDescription) {
    const task = this.tasks.get(taskId);
    if (!task) {
      console.log(`Task '${taskId}' not found.`);
      return false;
    }
    task.setDescription(newDescription);
    console.log(`Task '${taskId}' description updated.`);
    return true;
  }

  async updateTaskCategory(taskId, newCategory) {
    const task = this.tasks.get(taskId);
    if (!task) {
      console.log(`Task '${taskId}' not found.`);
      return false;
    }
    task.setCategory(newCategory);
    console.log(`Task '${taskId}' category updated to '${newCategory}'.`);
    return true;
  }

  async completeTask(taskId) {
    const task = this.tasks.get(taskId);
    if (!task) {
      console.log(`Task '${taskId}' not found.`);
      return false;
    }
    task.setCompleted(true);
    console.log(`Task '${taskId}' marked as completed.`);
    return true;
  }

  async uncompleteTask(taskId) {
    const task = this.tasks.get(taskId);
    if (!task) {
      console.log(`Task '${taskId}' not found.`);
      return false;
    }
    task.setCompleted(false);
    console.log(`Task '${taskId}' marked as pending.`);
    return true;
  }

  async deleteTask(taskId) {
    const task = this.tasks.get(taskId);
    if (!task) {
      console.log(`Task '${taskId}' not found.`);
      return false;
    }
    this.tasks.delete(taskId);

    const user = this.users.get(task.assignedUser);
    if (user) user.removeTaskId(taskId);

    console.log(`Task '${taskId}' deleted.`);
    return true;
  }

  getUserTasks(username) {
    const user = this.users.get(username);
    if (!user) return [];
    return user.taskIds.map(id => this.tasks.get(id)).filter(Boolean);
  }

  getTasksByCategory(category) {
    return [...this.tasks.values()].filter(
      task => task.category.toLowerCase() === category.toLowerCase()
    );
  }

  getPendingTasks(username) {
    return this.getUserTasks(username).filter(task => !task.completed);
  }

  getCompletedTasks(username) {
    return this.getUserTasks(username).filter(task => task.completed);
  }

  getAllTasks() {
    return [...this.tasks.values()];
  }

  getAllUsers() {
    return [...this.users.values()];
  }

  displayUserTasks(username) {
    const userTasks = this.getUserTasks(username);
    if (userTasks.length === 0) {
      console.log(`\nNo tasks found for user '${username}'.\n`);
      return;
    }
    console.log(`\n==================================================`);
    console.log(`Tasks for User: ${username}`);
    console.log(`==================================================`);
    userTasks.forEach(task => console.log(task.toDetailedString()));
  }

  displayStatistics() {
    const totalUsers = this.users.size;
    const totalTasks = this.tasks.size;
    const completed = [...this.tasks.values()].filter(t => t.completed).length;
    const pending = totalTasks - completed;

    console.log(`\n==================================================`);
    console.log(`System Statistics`);
    console.log(`==================================================`);
    console.log(`Total Users: ${totalUsers}`);
    console.log(`Total Tasks: ${totalTasks}`);
    console.log(`Completed Tasks: ${completed}`);
    console.log(`Pending Tasks: ${pending}`);
    console.log(`==================================================\n`);
  }
}

async function simulateUserAction(taskManager, username, description, category, delay) {
  await new Promise(res => setTimeout(res, delay));
  const task = await taskManager.createTask(username, description, category);
  if (task && Math.random() > 0.5) {
    await new Promise(res => setTimeout(res, delay / 2));
    await taskManager.completeTask(task.taskId);
  }
}

async function main() {
  console.log("==================================================");
  console.log("  Collaborative To-Do List Application (JavaScript)");
  console.log("==================================================\n");

  const taskManager = new TaskManager();

  console.log("PART 1: Basic CRUD Operations Demo\n");
  await taskManager.registerUser("Alice");
  await taskManager.registerUser("Bob");
  await taskManager.registerUser("Charlie");

  await taskManager.createTask("Alice", "Complete project proposal", "Work");
  await taskManager.createTask("Alice", "Review code changes", "Work");
  await taskManager.createTask("Alice", "Buy groceries", "Personal");
  await taskManager.createTask("Bob", "Prepare presentation", "Work");
  await taskManager.createTask("Bob", "Call dentist", "Personal");
  await taskManager.createTask("Charlie", "Fix bug in authentication", "Work");

  console.log("\nViewing Tasks");
  taskManager.displayUserTasks("Alice");

  console.log("\nUpdating Tasks");
  await taskManager.updateTaskDescription("TASK-0001", "Complete and submit project proposal");
  await taskManager.updateTaskCategory("TASK-0003", "Shopping");

  console.log("\nCompleting Tasks");
  await taskManager.completeTask("TASK-0001");
  await taskManager.completeTask("TASK-0004");

  console.log("\nViewing Updated Tasks");
  taskManager.displayUserTasks("Alice");

  console.log("\nDeleting a Task");
  await taskManager.deleteTask("TASK-0003");

  console.log("\nFinal Task List for Alice");
  taskManager.displayUserTasks("Alice");

  console.log("\nPART 2: Concurrent Operations Demo\n");

  const tasks = [
    simulateUserAction(taskManager, "Alice", "Implement user authentication", "Work", 500),
    simulateUserAction(taskManager, "Bob", "Write unit tests", "Work", 300),
    simulateUserAction(taskManager, "Charlie", "Update documentation", "Work", 400),
    simulateUserAction(taskManager, "Alice", "Schedule team meeting", "Work", 350),
    simulateUserAction(taskManager, "Bob", "Code review for PR #123", "Work", 250)
  ];

  await Promise.all(tasks);

  console.log("\nAll concurrent actions completed\n");

  console.log("\nTasks After Concurrent Updates");
  taskManager.displayUserTasks("Alice");
  taskManager.displayUserTasks("Bob");

  taskManager.displayStatistics();

  console.log("Application completed successfully.");
}

main();

File: CollaborativeToDoApp.java
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.stream.Collectors;

class Task {
    private final String taskId;
    private String description;
    private String category;
    private boolean completed;
    private final LocalDateTime createdAt;
    private LocalDateTime updatedAt;
    private final String assignedUser;

    public Task(String taskId, String description, String category, String assignedUser) {
        this.taskId = taskId;
        this.description = description;
        this.category = category;
        this.assignedUser = assignedUser;
        this.completed = false;
        this.createdAt = LocalDateTime.now();
        this.updatedAt = LocalDateTime.now();
    }

    public String getTaskId() { 
        return taskId; 
    }
    
    public String getDescription() { 
        return description; 
    }
    
    public String getCategory() { 
        return category; 
    }
    
    public boolean isCompleted() { 
        return completed; 
    }
    
    public LocalDateTime getCreatedAt() { 
        return createdAt; 
    }
    
    public LocalDateTime getUpdatedAt() { 
        return updatedAt; 
    }
    
    public String getAssignedUser() { 
        return assignedUser; 
    }

    public void setDescription(String description) {
        this.description = description;
        this.updatedAt = LocalDateTime.now();
    }

    public void setCategory(String category) {
        this.category = category;
        this.updatedAt = LocalDateTime.now();
    }

    public void setCompleted(boolean completed) {
        this.completed = completed;
        this.updatedAt = LocalDateTime.now();
    }

    public String getStatus() {
        return completed ? "Completed" : "Pending";
    }

    private String formatDateTime(LocalDateTime dateTime) {
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
        return dateTime.format(formatter);
    }

    public String toDetailedString() {
        StringBuilder sb = new StringBuilder();
        sb.append("Task ID: ").append(taskId).append("\n");
        sb.append("User: ").append(assignedUser).append("\n");
        sb.append("Description: ").append(description).append("\n");
        sb.append("Category: ").append(category).append("\n");
        sb.append("Status: ").append(getStatus()).append("\n");
        sb.append("Created: ").append(formatDateTime(createdAt)).append("\n");
        sb.append("Updated: ").append(formatDateTime(updatedAt)).append("\n");
        return sb.toString();
    }

    @Override
    public String toString() {
        return "Task[ID=" + taskId + ", User=" + assignedUser + 
               ", Description=" + description + ", Category=" + category + 
               ", Status=" + getStatus() + "]";
    }
}

class User {
    private final String username;
    private final List<String> taskIds;
    private final long createdAt;

    public User(String username) {
        this.username = username;
        this.taskIds = new ArrayList<>();
        this.createdAt = System.currentTimeMillis();
    }

    public String getUsername() {
        return username;
    }

    public synchronized List<String> getTaskIds() {
        return new ArrayList<>(taskIds);
    }

    public synchronized void addTaskId(String taskId) {
        if (!taskIds.contains(taskId)) {
            taskIds.add(taskId);
        }
    }

    public synchronized boolean removeTaskId(String taskId) {
        return taskIds.remove(taskId);
    }

    public synchronized int getTaskCount() {
        return taskIds.size();
    }

    public long getCreatedAt() {
        return createdAt;
    }

    public synchronized boolean hasTask(String taskId) {
        return taskIds.contains(taskId);
    }

    @Override
    public String toString() {
        return "User[username=" + username + ", tasks=" + taskIds.size() + "]";
    }
}

class TaskManager {
    private final ConcurrentHashMap<String, Task> tasks;
    private final ConcurrentHashMap<String, User> users;
    private final AtomicInteger taskCounter;

    public TaskManager() {
        this.tasks = new ConcurrentHashMap<>();
        this.users = new ConcurrentHashMap<>();
        this.taskCounter = new AtomicInteger(1);
    }

    public synchronized User registerUser(String username) {
        if (users.containsKey(username)) {
            System.out.println("User '" + username + "' already exists.");
            return null;
        }
        User user = new User(username);
        users.put(username, user);
        System.out.println("User '" + username + "' registered successfully.");
        return user;
    }

    public User getUser(String username) {
        return users.get(username);
    }

    public synchronized Task createTask(String username, String description, String category) {
        User user = users.get(username);
        if (user == null) {
            System.out.println("User '" + username + "' not found. Please register first.");
            return null;
        }

        String taskId = "TASK-" + String.format("%04d", taskCounter.getAndIncrement());
        Task task = new Task(taskId, description, category, username);
        tasks.put(taskId, task);
        user.addTaskId(taskId);

        System.out.println("Task '" + taskId + "' created for user '" + username + "'.");
        return task;
    }

    public Task getTask(String taskId) {
        return tasks.get(taskId);
    }

    public synchronized boolean updateTaskDescription(String taskId, String newDescription) {
        Task task = tasks.get(taskId);
        if (task == null) {
            System.out.println("Task '" + taskId + "' not found.");
            return false;
        }
        task.setDescription(newDescription);
        System.out.println("Task '" + taskId + "' description updated.");
        return true;
    }

    public synchronized boolean updateTaskCategory(String taskId, String newCategory) {
        Task task = tasks.get(taskId);
        if (task == null) {
            System.out.println("Task '" + taskId + "' not found.");
            return false;
        }
        task.setCategory(newCategory);
        System.out.println("Task '" + taskId + "' category updated to '" + newCategory + "'.");
        return true;
    }

    public synchronized boolean completeTask(String taskId) {
        Task task = tasks.get(taskId);
        if (task == null) {
            System.out.println("Task '" + taskId + "' not found.");
            return false;
        }
        task.setCompleted(true);
        System.out.println("Task '" + taskId + "' marked as completed.");
        return true;
    }

    public synchronized boolean uncompleteTask(String taskId) {
        Task task = tasks.get(taskId);
        if (task == null) {
            System.out.println("Task '" + taskId + "' not found.");
            return false;
        }
        task.setCompleted(false);
        System.out.println("Task '" + taskId + "' marked as pending.");
        return true;
    }

    public synchronized boolean deleteTask(String taskId) {
        Task task = tasks.remove(taskId);
        if (task == null) {
            System.out.println("Task '" + taskId + "' not found.");
            return false;
        }
        
        User user = users.get(task.getAssignedUser());
        if (user != null) {
            user.removeTaskId(taskId);
        }
        
        System.out.println("Task '" + taskId + "' deleted.");
        return true;
    }

    public List<Task> getUserTasks(String username) {
        User user = users.get(username);
        if (user == null) {
            return new ArrayList<>();
        }
        
        return user.getTaskIds().stream()
                .map(tasks::get)
                .filter(Objects::nonNull)
                .collect(Collectors.toList());
    }

    public List<Task> getTasksByCategory(String category) {
        return tasks.values().stream()
                .filter(task -> task.getCategory().equalsIgnoreCase(category))
                .collect(Collectors.toList());
    }

    public List<Task> getPendingTasks(String username) {
        return getUserTasks(username).stream()
                .filter(task -> !task.isCompleted())
                .collect(Collectors.toList());
    }

    public List<Task> getCompletedTasks(String username) {
        return getUserTasks(username).stream()
                .filter(Task::isCompleted)
                .collect(Collectors.toList());
    }

    public List<Task> getAllTasks() {
        return new ArrayList<>(tasks.values());
    }

    public List<User> getAllUsers() {
        return new ArrayList<>(users.values());
    }

    public void displayUserTasks(String username) {
        List<Task> userTasks = getUserTasks(username);
        
        if (userTasks.isEmpty()) {
            System.out.println("\nNo tasks found for user '" + username + "'.\n");
            return;
        }

        System.out.println("\n==================================================");
        System.out.println("Tasks for User: " + username);
        System.out.println("==================================================");
        
        for (Task task : userTasks) {
            System.out.println(task.toDetailedString());
        }
    }

    public void displayStatistics() {
        System.out.println("\n==================================================");
        System.out.println("System Statistics");
        System.out.println("==================================================");
        System.out.println("Total Users: " + users.size());
        System.out.println("Total Tasks: " + tasks.size());
        System.out.println("Completed Tasks: " + 
            tasks.values().stream().filter(Task::isCompleted).count());
        System.out.println("Pending Tasks: " + 
            tasks.values().stream().filter(task -> !task.isCompleted()).count());
        System.out.println("==================================================\n");
    }
}

class UserTaskWorker implements Runnable {
    private final TaskManager taskManager;
    private final String username;
    private final String description;
    private final String category;
    private final CountDownLatch latch;
    private final Random random;

    public UserTaskWorker(TaskManager taskManager, String username, 
                        String description, String category, CountDownLatch latch) {
        this.taskManager = taskManager;
        this.username = username;
        this.description = description;
        this.category = category;
        this.latch = latch;
        this.random = new Random();
    }

    @Override
    public void run() {
        try {
            Thread.sleep(random.nextInt(500));
            Task task = taskManager.createTask(username, description, category);
            Thread.sleep(random.nextInt(300));
            
            if (task != null && random.nextBoolean()) {
                Thread.sleep(random.nextInt(200));
                taskManager.completeTask(task.getTaskId());
            }
        } catch (InterruptedException e) {
            System.err.println("Thread interrupted: " + e.getMessage());
        } finally {
            latch.countDown();
        }
    }
}

public class CollaborativeToDoApp {
    
    public static void main(String[] args) {
        System.out.println("==================================================");
        System.out.println("  Collaborative To-Do List Application (Java)");
        System.out.println("==================================================\n");

        TaskManager taskManager = new TaskManager();

        System.out.println("PART 1: Basic CRUD Operations Demo\n");
        demonstrateBasicOperations(taskManager);

        System.out.println("\nPART 2: Concurrent Operations Demo\n");
        demonstrateConcurrency(taskManager);

        taskManager.displayStatistics();
        
        System.out.println("Application completed successfully.");
    }

    private static void demonstrateBasicOperations(TaskManager taskManager) {
        System.out.println("User Registration");
        taskManager.registerUser("Alice");
        taskManager.registerUser("Bob");
        taskManager.registerUser("Charlie");

        System.out.println("\nCreating Tasks");
        taskManager.createTask("Alice", "Complete project proposal", "Work");
        taskManager.createTask("Alice", "Review code changes", "Work");
        taskManager.createTask("Alice", "Buy groceries", "Personal");
        taskManager.createTask("Bob", "Prepare presentation", "Work");
        taskManager.createTask("Bob", "Call dentist", "Personal");
        taskManager.createTask("Charlie", "Fix bug in authentication", "Work");

        System.out.println("\nViewing Tasks");
        taskManager.displayUserTasks("Alice");

        System.out.println("\nUpdating Tasks");
        taskManager.updateTaskDescription("TASK-0001", "Complete and submit project proposal");
        taskManager.updateTaskCategory("TASK-0003", "Shopping");

        System.out.println("\nCompleting Tasks");
        taskManager.completeTask("TASK-0001");
        taskManager.completeTask("TASK-0004");

        System.out.println("\nViewing Updated Tasks");
        taskManager.displayUserTasks("Alice");

        System.out.println("\nDeleting a Task");
        taskManager.deleteTask("TASK-0003");

        System.out.println("\nFinal Task List for Alice");
        taskManager.displayUserTasks("Alice");
    }

    private static void demonstrateConcurrency(TaskManager taskManager) {
        System.out.println("Starting concurrent operations with 5 threads\n");

        ExecutorService executor = Executors.newFixedThreadPool(5);
        CountDownLatch latch = new CountDownLatch(5);

        String[][] userTasks = {
            {"Alice", "Implement user authentication", "Work"},
            {"Bob", "Write unit tests", "Work"},
            {"Charlie", "Update documentation", "Work"},
            {"Alice", "Schedule team meeting", "Work"},
            {"Bob", "Code review for PR #123", "Work"}
        };

        for (int i = 0; i < userTasks.length; i++) {
            final int index = i;
            executor.submit(new UserTaskWorker(
                taskManager, 
                userTasks[index][0], 
                userTasks[index][1], 
                userTasks[index][2],
                latch
            ));
        }

        try {
            latch.await();
            System.out.println("\nAll concurrent operations completed successfully\n");
        } catch (InterruptedException e) {
            System.err.println("Error waiting for threads: " + e.getMessage());
        }

        executor.shutdown();

        System.out.println("Demonstrating Concurrent Updates");
        demonstrateConcurrentUpdates(taskManager);
    }

    private static void demonstrateConcurrentUpdates(TaskManager taskManager) {
        ExecutorService executor = Executors.newFixedThreadPool(3);
        CountDownLatch latch = new CountDownLatch(3);

        executor.submit(() -> {
            try {
                Thread.sleep(100);
                taskManager.completeTask("TASK-0007");
                taskManager.completeTask("TASK-0009");
            } catch (InterruptedException e) {
                e.printStackTrace();
            } finally {
                latch.countDown();
            }
        });

        executor.submit(() -> {
            try {
                Thread.sleep(150);
                taskManager.updateTaskDescription("TASK-0008", "Write comprehensive unit tests");
            } catch (InterruptedException e) {
                e.printStackTrace();
            } finally {
                latch.countDown();
            }
        });

        executor.submit(() -> {
            try {
                Thread.sleep(200);
                taskManager.updateTaskCategory("TASK-0010", "Meeting");
            } catch (InterruptedException e) {
                e.printStackTrace();
            } finally {
                latch.countDown();
            }
        });

        try {
            latch.await();
            System.out.println("\nConcurrent updates completed\n");
        } catch (InterruptedException e) {
            System.err.println("Error: " + e.getMessage());
        }

        executor.shutdown();

        System.out.println("Tasks After Concurrent Updates");
        taskManager.displayUserTasks("Alice");
        taskManager.displayUserTasks("Bob");
    }
}